amida-tech/cql-execution
# Review 1st: HbA1c HEDIS Measure

# Related Tickets



- [SAR-53](https://jira.amida-tech.com/browse/SAR-53)

# Other Repos' PR(s) Intended to Work With This PR

# How Things Worked (or Didn't) Before This PR
James work was in a coffeescript format, which was using QI core and could only be used on FHIR 1.0.0 data (which is not what we'll most likely be handling)

# How Things Work Now (And How to Test)
## How it works now
I've taken James' coffeescript files and converted them into JS. Because we're using the JS version of the repo now, our QI Core CQL files needed to be converted into FHIR as well (since the JS version of the repo works in conjunction with a FHIR converter). I've also added a few other files that are helpful, like a new PR template, a determining coverage CQL file that will be used in all of our other CQL files at some point in the future, and a FHIR helpers file that allows us to use FHIR in the CQL files.

## To test
run `yarn` to install the necessary packages
run `node exec-files/exec-cdc_hba1c-lessThanEight.js`
Use your favorite diffchecker to confirm your output for `patientResults` and `unfilteredResults` matches what is below - you can ignore everything else.
```javascript
patientResults: {
    '96dcbd62-e4c6-a555-a663-77d08ad3c3b5': {
      Patient: [FHIRObject],
      'Initial Population': true,
      'Telehealth Modifier Encounter': true,
      'Telehealth POS Encounter': true,
      'Acute Inpatient Encounter': true,
      'Acute Inpatient Encounter Without Telehealth': false,
      'Acute Inpatient Discharge With Diabetes': [Array],
      'Claim with Acute Inpatient': [],
      'Acute Inpatient Discharge On Discharge Claim': false,
      'Nonacute Diabetes Discharges-List': [],
      'Nonacute Diabetes Discharges': 0,
      'Claim with Nonacute Inpatient Stay': [Array],
      'Nonacute Diabetes Discharges On Claim': 0,
      'Outpatient Visit': [],
      'Observation Visit': [],
      Out: [Array],
      'Telehealth Modifier With Diabetes': [Array],
      'Telephone Visit': [],
      'Online Assessment': [],
      'Out Tele Online with Different Dates': [],
      'Out Tele Online': 0,
      'At Least 2 Remote or Nonacute Inpatient Encounters With Diabetes': false,
      'Diabetes Diagnosis By Medication': false,
      Frailty: true,
      'Acute Inpatient With Advanced Illness': true,
      'Acute Inpatient Discharge With Advanced Illness': true,
      'Outpatient Advanced Illness': [],
      'Observation Advanced Illness': [],
      'ED Advanced Illness': [],
      'Nonacute AI': [],
      'Nonacute AI Discharge': [],
      'At Least Two Out/Obs/ED Advaned Illness': [],
      'Taking Dementia Medication': true,
      'Diabetes Exclusions': false,
      'Diabetes Condition': false,
      'Diabetes Condition Numerator': false,
      'Diabetes Condition Denominator': false,
      'HbA1c < 8': false,
      Numerator: 0,
      Denominator: 0
    },
    '440487f4-1137-f542-8127-dc98e392d21e': {
      Patient: [FHIRObject],
      'Initial Population': true,
      'Telehealth Modifier Encounter': true,
      'Telehealth POS Encounter': false,
      'Acute Inpatient Encounter': true,
      'Acute Inpatient Encounter Without Telehealth': false,
      'Acute Inpatient Discharge With Diabetes': [Array],
      'Claim with Acute Inpatient': [],
      'Acute Inpatient Discharge On Discharge Claim': false,
      'Nonacute Diabetes Discharges-List': [Array],
      'Nonacute Diabetes Discharges': 1,
      'Claim with Nonacute Inpatient Stay': [],
      'Nonacute Diabetes Discharges On Claim': 0,
      'Outpatient Visit': [],
      'Observation Visit': [],
      Out: [],
      'Telehealth Modifier With Diabetes': [Array],
      'Telephone Visit': [],
      'Online Assessment': [],
      'Out Tele Online with Different Dates': [],
      'Out Tele Online': 0,
      'At Least 2 Remote or Nonacute Inpatient Encounters With Diabetes': false,
      'Diabetes Diagnosis By Medication': true,
      Frailty: true,
      'Acute Inpatient With Advanced Illness': false,
      'Acute Inpatient Discharge With Advanced Illness': false,
      'Outpatient Advanced Illness': [],
      'Observation Advanced Illness': [],
      'ED Advanced Illness': [],
      'Nonacute AI': [],
      'Nonacute AI Discharge': [],
      'At Least Two Out/Obs/ED Advaned Illness': [],
      'Taking Dementia Medication': false,
      'Diabetes Exclusions': false,
      'Diabetes Condition': true,
      'Diabetes Condition Numerator': true,
      'Diabetes Condition Denominator': true,
      'HbA1c < 8': true,
      Numerator: 1,
      Denominator: 1
    }
  },
  unfilteredResults: {
    'Denominator Count': 1,
    'Numerator Count': 1,
    'CDC Measure Score': 1
  },
```
## To Review
Here's what to look for in each different file:
`exec-cdc_hba1c-lessThanEight.js` - This is the exec file that actually does the work - there's not much to review here.
`cdc_hba1c-codes.js` - The list of codes in each value set. This is programmatically generated using the CQL from the [VSAC](https://vsac.nlm.nih.gov/) API, so there's nothing to review here
`json-elm/FHIRHelpers.json` - Helper library that the repo creator told me I would need - nothing to review here
`cdc_hba1c-lessThanEight.json` - JSON version of the CQL created using the [clinical_quality_language](https://github.com/cqframework/clinical_quality_language) - nothing to review here
`cdc_hba1c-patients.js` - This contains the fake patients and their data. If you are really gung-ho, you can review the output of the results and confirm that the expressions with the CQL library returned the correct response based on the data, but that might take a while.
`cdc_hba1c-lessThanEight.cql` - this is the main piece of the PR. Read through the following requirements, and as best you can, confirm that the CQL matches the guidelines. If you'd prefer, whenever you get to this portion I can send you the word doc these requirements are in, which might be easier than looking at screenshots.
### NOTE: Per James (per Mike) The requirements list a bunch of medications, but apparently the valuesets for these medications are updated very regularly, so we need some way of programmatically getting these values which won't be covered in this ticket. James added a few medications as a demo
<img width="657" alt="Screen Shot 2020-12-09 at 12 34 03 PM" src="https://user-images.githubusercontent.com/22867407/105083002-58eaa680-5a62-11eb-93ce-7aee8cceeba2.png">
<img width="712" alt="Screen Shot 2020-12-09 at 12 34 18 PM" src="https://user-images.githubusercontent.com/22867407/105083014-5ab46a00-5a62-11eb-8634-42c2fd5e405e.png">
<img width="646" alt="Screen Shot 2020-12-09 at 12 34 25 PM" src="https://user-images.githubusercontent.com/22867407/105083021-5daf5a80-5a62-11eb-8337-f4b718bcce14.png">
<img width="641" alt="Screen Shot 2020-12-09 at 12 34 35 PM" src="https://user-images.githubusercontent.com/22867407/105083035-6142e180-5a62-11eb-9d4e-2f9f582cb82d.png">
<img width="669" alt="Screen Shot 2020-12-09 at 12 34 43 PM" src="https://user-images.githubusercontent.com/22867407/105083045-656eff00-5a62-11eb-9fd2-f7d144ceeb79.png">
<img width="718" alt="Screen Shot 2020-12-09 at 12 35 00 PM" src="https://user-images.githubusercontent.com/22867407/105083054-699b1c80-5a62-11eb-821c-19c638ab4232.png">
<img width="553" alt="Screen Shot 2020-12-10 at 12 40 03 PM" src="https://user-images.githubusercontent.com/22867407/105083074-6e5fd080-5a62-11eb-8f04-ab8eaed5882f.png">
